### ISSUE TITLE
Render NV2A commands using OpenGL

### ISSUE THREAD

**Issue by PatrickvL** [Cxbx-Reloaded/Cxbx-Reloaded#10]
Based on my research on the way the Xbox Direct3D API is translated to NV2A instructions, it's remarkable how closely these instruction match the OpenGL API. Many of the NV2A GPU's internal commands can be mapped directly to OpenGL calls, which makes OpenGL a more suitable rendering API than Direct3D.
(Also, using OpenGL as rendering API would be one less problem when porting to non-windows OS'es.)

**Comment by LukeUsher** [comment]
I agree, OpenGL is definitely the way to go. NVIDIA hardware seems to be designed around the spec, so it makes sense for us.

**Comment by PatrickvL** [comment]
The Dxbx project here on github, contains the beginnings of this exact idea. Search for USE_OPENGL in that codebase. 

One thing we must determine upfront, is the exact and correct layout of NV2A addresses.

Dxbx failed in this regard because the layout I used (which came from the nouveau project) got the projection matrix fields mixed up with the combined matrix. This resulted in garbage on screen instead of stable graphics, which wasone of the reasons I never continued on that avenue. 

We should really take this step by step, first making sure the basics work, before going on implementing more advanced stuff. Xdk samples and simple homebrew can go a long way in helping us with this incremental approach.

**Comment by PatrickvL** [comment]
Having experienced the hell that HLE emulation brings, going LLE might actually be preferrable.
Before this can start, we do need to implement a bit more of the kernel. As noted here https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/193#issuecomment-282669111 especially MmAllocateContiguousMemory is important, but https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/182 and https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/146 are probably needed too.

Next, learn as much as possible from https://github.com/espes/xqemu/blob/fix-fifo/hw/xbox/nv2a_int.h and https://github.com/espes/xqemu/blob/fix-fifo/hw/xbox/nv2a.c - perhaps it's even possible to make that code shareable between the two projects, benefitting both when improvements are done to it.

**Comment by haxar** [comment]
> Next, learn as much as possible from https://github.com/espes/xqemu/blob/fix-fifo/hw/xbox/nv2a_int.h and https://github.com/espes/xqemu/blob/fix-fifo/hw/xbox/nv2a.c - perhaps it's even possible to make that code shareable between the two projects, benefitting both when improvements are done to it.

I poked around in the Turok executable to see how NV2A commands were being generated. I took Turok because it was the first game to work with Cxbx, just enough to bootstrap HLE Xbox emulation:
http://xboxdevwiki.net/User:Haxar/NV2A

Since poking around in there, there are D3D8 variables that are being used internally, that are not being emulated, and cannot be emulated.

That led me to your article:
http://dxbx-emu.com/2010/12/02/better-emulation/

> Since I’ve been working in this code for the better part of 3 years already, I think I’m allowed to make a fairly bold statement here : We’re just doing it wrong.

Repurposing Cxbx/Dxbx to do NV2A, et al. hardware emulation would work. That would just leave all the kernel emulation.

The next challenge is to do NV2A hardware emulation on Linux, without a dependency on qemu. You will somehow need to map the GPU memory range 0xfd000000-0xfe000000 (size 0x01000000, 16MiB) and handle all R/W accesses individually, probably through a signal handler via SIGSEGV, and do dynamic patching, queue the register access to a pusher thread (xqemu), etc. The 16MiB range is segmented in 21 blocks as shown in xqemu's `nv2a_int.h` and `nv2a.c`.

But, my best bet is to access the NV2A registers directly via Xbox-Linux, and mmap() that range 1:1 virtual to physical. Emulation would be better on an Xbox than a typical PC. :P Just to get started.
